src/core/mempool.rs

```diff
@@ -18,18 +65,333 @@ pub struct MempoolAdmitter {
+
+        if create_flag {
+            // instantiate!
+            MemPoolDB::instantiate_mempool_db(&mut conn)?;
+        }
+        
+        Ok(MemPoolDB {
+            db: conn,
+            path: db_path.to_string(),
+            admitter: admitter,
+        })
+    }
+
+    pub fn conn(&self) -> &DBConn {
+        &self.db
+    }
+
+    pub fn tx_begin<'a>(&'a mut self) -> Result<MemPoolTx<'a>, db_error> {
+        let tx = self.db.transaction().map_err(db_error::SqliteError)?;
+        Ok(MemPoolTx::new(tx, &mut self.admitter))
+    }
+
+    fn db_has_tx(conn: &DBConn, txid: &Txid) -> Result<bool, db_error> {
+        query_row(conn, "SELECT 1 FROM mempool WHERE txid = ?1", &[txid as &dyn ToSql])
+            .and_then(|row_opt: Option<i64>| Ok(row_opt.is_some()))
+    }
+
+    pub fn get_tx(conn: &DBConn, txid: &Txid) -> Result<Option<MemPoolTxInfo>, db_error> {
+        query_row(conn, "SELECT * FROM mempool WHERE txid = ?1", &[txid as &dyn ToSql])
+    }
+    
+    fn get_tx_fee(conn: &DBConn, txid: &Txid) -> Result<Option<u64>, db_error> {
+        query_row(conn, "SELECT fee FROM mempool WHERE txid = ?1", &[txid as &dyn ToSql])
+    }
+
+    /// Get all transactions across all tips
+    #[cfg(test)]
+    pub fn get_all_txs(conn: &DBConn) -> Result<Vec<MemPoolTxInfo>, db_error> {
+        let sql = "SELECT * FROM mempool";
+        let rows = query_rows::<MemPoolTxInfo, _>(conn, &sql, NO_PARAMS)?;
+        Ok(rows)
+    }
+
+    /// Get a number of transactions after a given timestamp on a given chain tip.
+    /// Order the resulting transactions by fee, in decreasing order.
+    pub fn get_txs_after(conn: &DBConn, burn_header_hash: &BurnchainHeaderHash, block_header_hash: &BlockHeaderHash, timestamp: u64, count: u64) -> Result<Vec<MemPoolTxInfo>, db_error> {
+        let sql = "SELECT * FROM mempool WHERE accept_time >= ?1 AND burn_header_hash = ?2 AND block_header_hash = ?3 ORDER BY fee DESC LIMIT ?4";
+        let args : &[&dyn ToSql] = &[&u64_to_sql(timestamp)?, burn_header_hash, block_header_hash, &u64_to_sql(count)?];
+        let rows = query_rows::<MemPoolTxInfo, _>(conn, &sql, args)?;
+        Ok(rows)
+    }
+
+    /// Add a transaction to the mempool.  If it already exists, then replace it if the given fee
+    /// is higher than the one that's already there.
+    /// Carry out the mempool admission test before adding.
+    /// NOTE: fee is the _total fee_ of the transaction, not the fee rate!
+    /// Don't call directly; use submit()
+    fn try_add_tx<'a>(tx: &mut MemPoolTx<'a>, burn_header_hash: &BurnchainHeaderHash, block_header_hash: &BlockHeaderHash, txid: Txid, tx_bytes: Vec<u8>, fee: u64, nonce: u64, height: u64) -> Result<(), MemPoolRejection> {
+        let len = tx_bytes.len() as u64;
+
+        // replace-by-fee
+        if let Some(tx_fee) = MemPoolDB::get_tx_fee(&tx, &txid).map_err(MemPoolRejection::DBError)? {
```

Comment:
Is this possibly true? I think the fee is included in the txid, so we could never (modulo hash collision) have a transaction with the same txid but a different fee, right?

Reply:
D'oh! You're right.  What should be happening here is the mempool should be storing the origin and spending accounts' addresses and nonces, and looking to see if a _different_ transaction arrives with the _same_ addresses and nonces (in which case, the transaction would be replaced by the one with a higher fee).  Will fix.

Reply:
Fixed in d27c54d.  Transactions are keyed by `(origin_address, origin_nonce, sponsor_address, sponsor_nonce)` now, so they can be replaced by a different transaction with a higher fee while still unconfirmed.